spring-cloud-starter-stream-source-sftp/src/main/java/org/springframework/cloud/stream/app/sftp/source/tasklauncher/SftpTaskLaunchRequestContextProvider.java

```diff
@@ -0,0 +1,224 @@
+import org.springframework.cloud.stream.app.sftp.source.ListFilesRotator;
+import org.springframework.cloud.stream.app.sftp.source.SftpHeaders;
+import org.springframework.cloud.stream.app.sftp.source.SftpSourceProperties;
+import org.springframework.cloud.stream.app.sftp.source.task.SftpSourceTaskProperties;
+import org.springframework.cloud.stream.app.tasklaunchrequest.TaskLaunchRequestContext;
+import org.springframework.cloud.stream.app.tasklaunchrequest.TaskLaunchRequestType;
+import org.springframework.cloud.stream.app.tasklaunchrequest.TaskLaunchRequestTypeProvider;
+import org.springframework.integration.expression.FunctionExpression;
+import org.springframework.integration.handler.MessageProcessor;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Provide {@link org.springframework.cloud.stream.app.tasklaunchrequest.TaskLaunchRequestContext}, specific to
+ * SFTP. as a message header.
+ *
+ * @author David Turanski
+ **/
+
+public class SftpTaskLaunchRequestContextProvider implements MessageProcessor<Message> {
+	private final static Log log = LogFactory.getLog(SftpTaskLaunchRequestContextProvider.class);
+
+	private final SftpSourceTaskProperties sftpSourceTaskProperties;
+	private final SftpSourceProperties sourceProperties;
+	private final LocalDirectoryResolver localDirectoryResolver;
+	private final TaskLaunchRequestTypeProvider taskLaunchRequestTypeProvider;
+	private final ListFilesRotator listFilesRotator;
+
+	public SftpTaskLaunchRequestContextProvider(
+		SftpSourceTaskProperties sftpSourceTaskProperties, SftpSourceProperties sourceProperties,
+		LocalDirectoryResolver localDirectoryResolver, TaskLaunchRequestTypeProvider taskLaunchRequestTypeProvider,
+		ListFilesRotator listFilesRotator) {
+		Assert.notNull(sftpSourceTaskProperties, "'sftpSourceTaskProperties' is required");
+		Assert.notNull(sourceProperties, "'sourceProperties' is required");
+		Assert.notNull(localDirectoryResolver, "'localDirectoryResolver' is required");
+		Assert.notNull(taskLaunchRequestTypeProvider, "'taskLaunchRequestTypeProvider' is required");
+
+		this.sftpSourceTaskProperties = sftpSourceTaskProperties;
+		this.sourceProperties = sourceProperties;
+		this.localDirectoryResolver = localDirectoryResolver;
+		this.taskLaunchRequestTypeProvider = taskLaunchRequestTypeProvider;
+		this.listFilesRotator = listFilesRotator;
+	}
+
+	@Override
+	public Message<?> processMessage(Message<?> message) {
+		if (taskLaunchRequestTypeProvider.taskLaunchRequestType() == TaskLaunchRequestType.NONE) {
+			return message;
+		}
+
+		log.debug(String.format("Preparing context for a %s task launch request", taskLaunchRequestTypeProvider
+			.taskLaunchRequestType().name()));
+
+		if (listFilesRotator != null) {
+			message = MessageBuilder.fromMessage(message).copyHeaders(listFilesRotator.headers()).build();
```

Comment:
Would be better do not build the whole message here, but just have a `AbstractMessageBuilder` result and build message only in the end of the process. Otherwise you create two messages and have some overhead with `ID` and `TIMESTAMP` headers generation.